### ISSUE TITLE
Snapshots

### ISSUE THREAD

**Issue by mrjoshuak** [absfs/boltfs#3]
Starting this issue to track the snapshots topic initially discussed in #2. 

Currently there is no snapshot functionality in any of the absfs implementations. Perhaps boltfs is a good candidate for a first implementation.  I'm guessing this would need to be some sort of MVCC implementation of inodes, and immutable file contents (aka copy on write). 

Copy on write can be implemented as it's own absfs implementation so that it can be composed with any other absfs implementation, but filesystem implementation details such as inodes are intentionally not exposed by the absfs interface so an MVCC enable implementation would be most appropriate in either the individual file system implementations, or perhaps in the [inode](https://github.com/absfs/inode).

We're in need of some design guidance here, so that we can come up with an implementation plan. 

I'll try some things out, but in the meantime everyone is welcome to add comments and suggestions regarding snapshot implementation to this thread.

**Comment by ismailkuru** [comment]
I agree that boltfs is a good candidate based on the fact that the underlying store "boltdb" is CoW. Any update to an inode in boltfs allocates and write all pages touched by the update.

So taking a snapshot at this point is not that difficult -- or list of snapshots. Here is my scratch for a potential snapshot taking:

- Starting from the root directory of the boltfs
snapshot{
            ....
            snapshotcursor := rootbtree.Directories().List(rootbtree.inode)
    
           for snapshotcursor.First(), ,snapshotcursor.Next()
           {
                  case:
                          entry_is_directory then child.snapshot()
                          entry is file then add entry to the buffer to be written 
            }
}
where we can keep the list of directories in the "state" btree of boltfs.

- We can write them to ".snap" file and read them.
- We can also keep a list of snapshots --- periodically write one of them as a checkpoint and drop the rest of snapshots.

**Comment by ismailkuru** [comment]
One another option would be lifting the tx.WriteTo into the inode level.

**Comment by mrjoshuak** [comment]
Db level copy on write and filesystem level copy on write are different.  I'm not sure how one should handle snapshots taken when files are open for writing. While the db snapshot would be invariant during the snapshot, open files could cause the snapshot to capture the filesystem in an invalid state. I.e. neither the file before modification, or the file after modification, but some partial state. This is made worse if the file data is written to storage outside of the database itself. 

I think we have to have a transactional filesystem implementation in place so that snapshots can be attached to a concrete filesystem state, and in progress filesystem transactions do not become part of the snapshot.

**Comment by mrjoshuak** [comment]
This article illustrates this issue somewhat.  In Jose Faleiro's article [Mvcc Isolation](http://www.jmfaleiro.com/blog/post/mvcc-isolation/), he refers to it as serializable isolation.

**Comment by ismailkuru** [comment]
From my point of view, each system call has to be implemented inside a transaction. This will allow you to utilize the underlying write-methodology. To be more clear, instead of simulating the filesystem on top of db, boltfs should project the underlying storage-DB as a filesystem. How could a file be represented on the underlying DB or a directory etc. ? After a clear identification of these issues, we can utilize the underlying write-methodology of storage system?

**Comment by ismailkuru** [comment]
Of course, this requires changes to the layout of the boltfs. Currently, it bookkeeps different parts of the filesystem in different buckets. Also, mvcc is so strong for a filesystem, and it will hurt the performance of the filesystem a lot. So, the underlying db also requires a bit of change in the commit of a transaction where instead of keeping txn local page caches, there should be global cache shared among different txns (readers should be able to look at the page cache if it is not accessed by the writer exclusively).

**Comment by mrjoshuak** [comment]
I'm considering a design in which we hash the file contents, and inodes.  Inodes include the hash of file contents so the hash of an inode represents both the filesystem entity and the data.   Directory inodes would contain the Merkle root of its hashed members. A filesystem mutating transaction starts by noting the hash of any directory below which a mutation occurs within the transaction, when the transaction is committed the directory hashes are checked from the root down.  If the root directory still contains the noted hash a compare and swap is performed using the database transaction as the CAS mechanism. The transaction can be replayed if a compare fails, and as an optimization only the portions of the transaction that affect directories with updated hashes need to be replayed.  In other words conflicting writes are not actually conflicting if they touch different directories. 

Open file handles can be treated as an entirely separate context that is isolated from filesystem consistency.  When a file that already exists is opened, a reader is directed to the file contents via the hash of the content and so will not loose access to the data even if the file is removed in the mean time.  Writing to a file can be handled entirely outside of any transactions, or within the context of a filesystem transaction, but in neither of these cases is the database or filesystem consistency directly effected since nothing is actually committed until the file is closed and the filesystem transaction successfully commits. Opening a file in one thread for reading and in another for writing such as with named pipes, might even be supportable as a special case since writing happens outside of the consistency layers. 

In short, boltdb need only provide snapshot isolation, since arbitrarily complex filesystem transactions can be reduced to a single CAS operation on the root inode. Serializability is provided by the filesystem transaction, and active file handles are managed independently. 

This is a change in the current architecture, which currently allows inodes to mutate, but it seems doable. Importantly it doesn't require any change or special knowledge of boltdb internals though it does leverage boltdb for isolation that for the most part other backends don't provide.  I don't believe there are any blocking / locking mechanics required. There may be some complexity around garbage collection, since there are at least two notions of "unreachable" inodes.  Inodes that cannot be reached in the current filesystem view (but could in some previous view) , and inodes that were created but subsequently failed a CAS operation and therefore are not valid in any view.

**Comment by ismailkuru** [comment]
> I'm considering a design in which we hash the file contents, and inodes. Inodes include the hash of file contents so the hash of an inode represents both the filesystem entity and the data. Directory inodes would contain the Merkle root of its hashed members. A filesystem mutating transaction starts by noting the hash of any directory below which a mutation occurs within the transaction, when the transaction is committed the directory hashes are checked from the root down. If the root directory still contains the noted hash a compare and swap is performed using the database transaction as the CAS mechanism. The transaction can be replayed if a compare fails, and as an optimization only the portions of the transaction that affect directories with updated hashes need to be replayed. In other words conflicting writes are not actually conflicting if they touch different directories.

--- So you still assume the entire content of the file in the memory. 

> Open file handles can be treated as an entirely separate context that is isolated from filesystem consistency. When a file that already exists is opened, a reader is directed to the file contents via the hash of the content and so will not loose access to the data even if the file is removed in the mean time. Writing to a file can be handled entirely outside of any transactions, or within the context of a filesystem transaction, but in neither of these cases is the database or filesystem consistency directly effected since nothing is actually committed until the file is closed and the filesystem transaction successfully commits. Opening a file in one thread for reading and in another for writing such as with named pipes, might even be supportable as a special case since writing happens outside of the consistency layers.
> 
> In short, boltdb need only provide snapshot isolation, since arbitrarily complex filesystem transactions can be reduced to a single CAS operation on the root inode. Serializability is provided by the filesystem transaction, and active file handles are managed independently.
> 
> This is a change in the current architecture, which currently allows inodes to mutate, but it seems doable. Importantly it doesn't require any change or special knowledge of boltdb internals though it does leverage boltdb for isolation that for the most part other backends don't provide. 

--- I agree that the changes offered do not require boltdb knowledge.

I don't believe there are any blocking / locking mechanics required. There may be some complexity around garbage collection, since there are at least two notions of "unreachable" inodes. Inodes that cannot be reached in the current filesystem view (but could in some previous view) , and inodes that were created but subsequently failed a CAS operation and therefore are not valid in any view.

**Comment by mrjoshuak** [comment]
We can support file contents in the database, or on disk or both. I don't believe files have to be stored entirely in memory in either case. Using a file hash allows us to completely deffer file data handeling to any system at all and only have to deal with the filesystem internals.  The hash can be computed as a file is streamed to storage, regardless of where that actually is. Importantly the data and inodes can be written to the db completely in parallel since there is no possibility of conflict only the root inode must be atomically replaced.

**Comment by mrjoshuak** [comment]
Implemented as of [275fd97](https://github.com/absfs/boltfs/commit/275fd97627d472259cb5f66da8fb1eb942a44b20)
